readthedocs/restapi/views/integrations.py

```diff
@@ -0,0 +1,172 @@
+
+            {
+                "build_triggered": true,
+                "project": "project_name",
+                "versions": [...]
+            }
+        """
+        raise NotImplementedError
+
+
+class GitHubWebhookView(WebhookMixin, APIView):
+
+    """Webhook consumer for GitHub
+
+    Accepts webhook events from GitHub, 'push' events trigger builds. Expects the
+    webhook event type will be included in HTTP header ``X-GitHub-Event``, and
+    we will have a JSON payload.
+
+    Expects the following JSON::
+
+        {
+            "ref": "branch-name",
+            ...
+        }
+
+    If a build is triggered, this will return JSON with the following::
+
+        {
+            "build_triggered": true,
+            "project": "project_name",
+            "versions": [...]
+        }
+    """
+
+    def handle_webhook(self, request, project, data=None):
+        # Get event and trigger other webhook events
+        event = request.META.get('HTTP_X_GITHUB_EVENT', 'push')
+        webhook_github.send(Project, project=project, data=data, event=event)
+        # Handle push events and trigger builds
+        try:
+            branches = [request.data['ref'].replace('refs/heads/', '')]
+        except KeyError:
+            raise ParseError('Parameter "ref" is required')
+        if event == GITHUB_PUSH:
+            to_build, not_building = build_branches(project, branches)
+            triggered = True if to_build else False
+            return {'build_triggered': triggered,
+                    'project': project.slug,
+                    'versions': to_build}
+        return {'build_triggered': False}
+
+
+class GitLabWebhookView(WebhookMixin, APIView):
+
+    """Webhook consumer for GitLab
+
+    Accepts webhook events from GitLab, 'push' events trigger builds.
+
+    Expects the following JSON::
+
+        {
+            "object_kind": "push",
+            "ref": "branch-name",
+            ...
+        }
+
+    If a build is triggered, this will return JSON with the following::
+
+        {
+            "build_triggered": true,
+            "project": "project_name",
+            "versions": [...]
+        }
+    """
+
+    def handle_webhook(self, request, project, data=None):
+        # Get event and trigger other webhook events
+        event = data.get('object_kind', GITLAB_PUSH)
+        webhook_gitlab.send(Project, project=project, data=data, event=event)
+        # Handle push events and trigger builds
+        try:
+            branches = [request.data['ref'].replace('refs/heads/', '')]
+        except KeyError:
+            raise ParseError('Parameter "ref" is required')
+        if event == GITLAB_PUSH:
+            to_build, not_building = build_branches(project, branches)
+            triggered = True if to_build else False
+            return {'build_triggered': triggered,
+                    'project': project.slug,
+                    'versions': to_build}
+        return {'build_triggered': False}
```

Comment:
This code feels like it could be abstracted, and it's the same for GH/BB except for a couple lines.